[core ATen IR] Add decompositions for max, min, var_mean

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110906

## Context

Add decompositions for `aten.max`, `aten.min`, and `aten.var_mean`. These operators follow a pattern of returning a tuple of outputs from two component operators:

```
aten.max(x) -> return aten.amax(x), aten.argmax(x)
aten.min(x) -> return aten.amin(x), aten.argmin(x)
aten.var_mean(x) -> return aten.var(x), aten.mean(x)
```

For `var_mean`, the `refs` implementation was doing something similar, so I changed it to call `torch.` ops instead like was done for other `refs` implementations previously. cc: @peterbell10 @lezcano 

Note that Inductor lowers all these directly, so they are excluded from the Inductor decomp table.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan

SS-JIA committed:
[core ATen IR] Add decompositions for max, min, var_mean

[ghstack-poisoned]

SS-JIA committed:
Update on "[core ATen IR] Add decompositions for max, min, var_mean"


## Context

Add decompositions for `aten.max`, `aten.min`, and `aten.var_mean`. These operators follow a pattern of returning a tuple of outputs from two component operators:

```
aten.max(x) -> return aten.amax(x), aten.argmax(x)
aten.min(x) -> return aten.amin(x), aten.argmin(x)
aten.var_mean(x) -> return aten.var(x), aten.mean(x)
```

For `var_mean`, the `refs` implementation was doing something similar, so I changed it to call `torch.` ops instead like was done for other `refs` implementations previously. cc: peterbell10 lezcano 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

SS-JIA commented on test/test_decomp.py:
```diff
@@ -699,13 +709,14 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
-                 f"CompositeImplicitAutograd you should skip this test "
-                 f"by updating CROSS_REF_EXCLUDE_SET.")
-        )
+        if aten_name not in SKIP_CHECK_DECOMPOSED:
```
@peterbell10 Previously we had avoided doing this, but since there is no `refs` implementation for `max` and `min` I couldn't figure out a better way to get around this. let me know if there are any other options.

SS-JIA replied:
I could move these decomps to refs if that is your preference,but this problem will still persist. I think the issue is that for the max_binary test cases, calling e.g. torch.max will redirect to aten.maximum under the hood. So the test will see that nothing gets decomposed.

SS-JIA replied:
From a conceptual standpoint though, I think a reference implementation of this op should be a one pass implementation. So my personal preference is to leave it as-is.

peterbell10 replied:
> I could move these decomps to refs if that is your preference,but this problem will still persist.

Oh I see, the issue here is that  `max.other` and `max.out` are `CompositeImplicitAutograd` so the decomposition is never hit. In that case the fix should be to remove their decompositions.

> From a conceptual standpoint though, I think a reference implementation of this op should be a one pass implementation. So my personal preference is to leave it as-is.

The reference for `var_mean` decomposes into var + mean so I don't think there's any such rule.

SS-JIA replied:
> Oh I see, the issue here is that max.other and max.out are CompositeImplicitAutograd so the decomposition is never hit. In that case the fix should be to remove their decompositions.

Even with the decompositions removed, `max.other` and `max.out` should be redirected to `maximum`, is that correct? I'm not familiar with what `CompositeImplicitAutograd` means, unfortunately.

peterbell10 replied:
Yes, `CompositeImplicitAutograd` operators are transparent to the tracing mechanism so `maximum` will be traced directly without `ProxyTensor` ever getting dispatched for `max`.

SS-JIA committed:
Update on "[core ATen IR] Add decompositions for max, min, var_mean"


## Context

Add decompositions for `aten.max`, `aten.min`, and `aten.var_mean`. These operators follow a pattern of returning a tuple of outputs from two component operators:

```
aten.max(x) -> return aten.amax(x), aten.argmax(x)
aten.min(x) -> return aten.amin(x), aten.argmin(x)
aten.var_mean(x) -> return aten.var(x), aten.mean(x)
```

For `var_mean`, the `refs` implementation was doing something similar, so I changed it to call `torch.` ops instead like was done for other `refs` implementations previously. cc: peterbell10 lezcano 

Note that Inductor lowers all these directly, so they are excluded from the Inductor decomp table.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

SS-JIA committed:
Update on "[core ATen IR] Add decompositions for max, min, var_mean"


## Context

Add decompositions for `aten.max`, `aten.min`, and `aten.var_mean`. These operators follow a pattern of returning a tuple of outputs from two component operators:

```
aten.max(x) -> return aten.amax(x), aten.argmax(x)
aten.min(x) -> return aten.amin(x), aten.argmin(x)
aten.var_mean(x) -> return aten.var(x), aten.mean(x)
```

For `var_mean`, the `refs` implementation was doing something similar, so I changed it to call `torch.` ops instead like was done for other `refs` implementations previously. cc: peterbell10 lezcano 

Note that Inductor lowers all these directly, so they are excluded from the Inductor decomp table.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

SS-JIA committed:
Update on "[core ATen IR] Add decompositions for max, min, var_mean"


## Context

Add decompositions for `aten.max`, `aten.min`, and `aten.var_mean`. These operators follow a pattern of returning a tuple of outputs from two component operators:

```
aten.max(x) -> return aten.amax(x), aten.argmax(x)
aten.min(x) -> return aten.amin(x), aten.argmin(x)
aten.var_mean(x) -> return aten.var(x), aten.mean(x)
```

For `var_mean`, the `refs` implementation was doing something similar, so I changed it to call `torch.` ops instead like was done for other `refs` implementations previously. cc: peterbell10 lezcano 

Note that Inductor lowers all these directly, so they are excluded from the Inductor decomp table.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan

[ghstack-poisoned]

SS-JIA commented on torch/_decomp/__init__.py:
```diff
@@ -342,6 +342,18 @@ def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
             aten.logsumexp.default,
             aten.masked_fill,
             aten.masked_fill_,
+            aten.max.default,
+            aten.max.dim,
+            aten.max.dim_max,
+            aten.max.other,
+            aten.max.out,
+            aten.max.unary_out,
+            aten.min.default,
+            aten.min.dim,
+            aten.min.dim_min,
+            aten.min.other,
+            aten.min.out,
+            aten.min.unary_out,
```
I'll try that. I didn't hit all the available overloads of this op so I wasn't sure if it would work 😅

SS-JIA replied:
Fixed (y)

SS-JIA committed:
Update on "[core ATen IR] Add decompositions for max, min, var_mean"


## Context

Add decompositions for `aten.max`, `aten.min`, and `aten.var_mean`. These operators follow a pattern of returning a tuple of outputs from two component operators:

```
aten.max(x) -> return aten.amax(x), aten.argmax(x)
aten.min(x) -> return aten.amin(x), aten.argmin(x)
aten.var_mean(x) -> return aten.var(x), aten.mean(x)
```

For `var_mean`, the `refs` implementation was doing something similar, so I changed it to call `torch.` ops instead like was done for other `refs` implementations previously. cc: peterbell10 lezcano 

Note that Inductor lowers all these directly, so they are excluded from the Inductor decomp table.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan

[ghstack-poisoned]

SS-JIA committed:
Update on "[core ATen IR] Add decompositions for max, min, var_mean"


## Context

Add decompositions for `aten.max`, `aten.min`, and `aten.var_mean`. These operators follow a pattern of returning a tuple of outputs from two component operators:

```
aten.max(x) -> return aten.amax(x), aten.argmax(x)
aten.min(x) -> return aten.amin(x), aten.argmin(x)
aten.var_mean(x) -> return aten.var(x), aten.mean(x)
```

For `var_mean`, the `refs` implementation was doing something similar, so I changed it to call `torch.` ops instead like was done for other `refs` implementations previously. cc: peterbell10 lezcano 

Note that Inductor lowers all these directly, so they are excluded from the Inductor decomp table.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan

[ghstack-poisoned]

SS-JIA commented on torch/_refs/__init__.py:
```diff
@@ -2396,6 +2396,100 @@ def amax(
+@register_decomposition(
+    [aten.max.default, aten.max.unary_out, aten.max.dim, aten.max.dim_max]
+)
+def unary_max(
```
@peterbell10 moved these to `_refs` as suggested.

peterbell10 replied:
Can you fix the names to match the python api and add `PythonRefInfo` entries so we get the tests running on these? 

https://github.com/pytorch/pytorch/blob/5183760ca52ea7dbd892ebc88cabdd43ebc81cbe/torch/testing/_internal/common_methods_invocations.py#L21411

You will also need multiple entries with different `torch_opinfo_variant_name` values to pick up all the eager opinfos.

SS-JIA replied:
Sure thing.

> You will also need multiple entries with different torch_opinfo_variant_name values to pick up all the eager opinfos.

Does this refer to multiple `OpInfo` entries such as [this](https://fburl.com/ouqgzol1)?

peterbell10 replied:
I can't see your meta-internal link but this is an example where an operator has multiple opinfos:

https://github.com/pytorch/pytorch/blob/5183760ca52ea7dbd892ebc88cabdd43ebc81cbe/torch/testing/_internal/common_methods_invocations.py#L19558-L19575

SS-JIA replied:
@peterbell10 so running into some slight issues with renaming `unary_max` to `max`. There already is an [existing refs implementation for `maximum`](https://github.com/pytorch/pytorch/blob/5183760ca52ea7dbd892ebc88cabdd43ebc81cbe/torch/_refs/__init__.py#L1641-L1642). If I rename `unary_max` to `max`, then the test cases for `maximum` will try to call `_refs.max` instead of `_refs.maximum`, which will fail because the function signatures are completely different. I'm not sure why this happens.

One workaround I can think of is to keep the `unary_max` name for the function, and add `PythonRefInfo` entries registered to `_refs.unary_max` accordingly. Is this an acceptable solution?

SS-JIA replied:
Hmm, now running into issues with `torch.argmax` not having a refs implementation, so the tests for `unary_max` will fail as well.

In light of these challenges, I think it may be more appropriate to just move these decomps back to `decompositions.py`, since we are unable to test the decomposition as a reference implementation. What do you think?

SS-JIA committed:
Update on "[core ATen IR] Add decompositions for max, min, var_mean"


## Context

Add decompositions for `aten.max`, `aten.min`, and `aten.var_mean`. These operators follow a pattern of returning a tuple of outputs from two component operators:

```
aten.max(x) -> return aten.amax(x), aten.argmax(x)
aten.min(x) -> return aten.amin(x), aten.argmin(x)
aten.var_mean(x) -> return aten.var(x), aten.mean(x)
```

For `var_mean`, the `refs` implementation was doing something similar, so I changed it to call `torch.` ops instead like was done for other `refs` implementations previously. cc: peterbell10 lezcano 

Note that Inductor lowers all these directly, so they are excluded from the Inductor decomp table.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan

[ghstack-poisoned]

peterbell10 commented on test/test_decomp.py:
```diff
@@ -431,6 +431,11 @@ def test_unsupported(t):
     return any(test_unsupported(x) for x in itertools.chain(flat_args, flat_kwargs))
 
 
+core_not_decomposed = {
```
No you shouldn't need to skip any tests. Just don't register the aliases to maximum as decompositions.

peterbell10 replied:
Or are the tests still failing now?

SS-JIA replied:
Yes, if I don't register the decompositions, the tests will still fail. But it makes sense since no decomposition occurs. I think the only way to prevent it is to skip the tests.

SS-JIA replied:
My understanding is that since even though `max.other` doesn't have a decomposition registered, it will still be tested since other variants of `max` have decompositions. If `max.other` is tested, then `check_decomposed()` will fail since `maximum` gets called instead.

peterbell10 replied:
What error do you get?

SS-JIA replied:
```
AssertionError: False is not true : aten.max was not decomposed, saw calls for: aten.maximum.default. If your op is  CompositeImplicitAutograd you should skip this test by updating CROSS_REF_EXCLUDE_SET.

To execute this test, run the following from the base repo dir:
     python test/test_decomp.py -k test_quick_max_binary_cpu_uint8
```

peterbell10 replied:
Hrm okay, I was expecting it to say `aten.max.other` but if it isn't checking at the overload level then I guess we do need to skip.

SS-JIA committed:
Update on "[core ATen IR] Add decompositions for max, min, var_mean"


## Context

Add decompositions for `aten.max`, `aten.min`, and `aten.var_mean`. These operators follow a pattern of returning a tuple of outputs from two component operators:

```
aten.max(x) -> return aten.amax(x), aten.argmax(x)
aten.min(x) -> return aten.amin(x), aten.argmin(x)
aten.var_mean(x) -> return aten.var(x), aten.mean(x)
```

For `var_mean`, the `refs` implementation was doing something similar, so I changed it to call `torch.` ops instead like was done for other `refs` implementations previously. cc: peterbell10 lezcano 

Note that Inductor lowers all these directly, so they are excluded from the Inductor decomp table.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan

[ghstack-poisoned]

SS-JIA commented on torch/_decomp/decompositions.py:
```diff
@@ -4232,6 +4232,100 @@ def squeeze_default(self: Tensor, dim: Optional[int] = None):
+@register_decomposition(
+    [aten.max.default, aten.max.unary_out, aten.max.dim, aten.max.dim_max]
+)
+def unary_max(
```
moving these back to `decompositions.py`. Copy pasting from the other comment thread:

> @peterbell10 so running into some slight issues with renaming unary_max to max. There already is an [existing refs implementation for maximum](https://github.com/pytorch/pytorch/blob/5183760ca52ea7dbd892ebc88cabdd43ebc81cbe/torch/_refs/__init__.py#L1641-L1642). If I rename unary_max to max, then the test cases for maximum will try to call _refs.max instead of _refs.maximum, which will fail because the function signatures are completely different. I'm not sure why this happens.

> One workaround I can think of is to keep the unary_max name for the function, and add PythonRefInfo entries registered to _refs.unary_max accordingly. Is this an acceptable solution?

> Hmm, now running into issues with torch.argmax not having a refs implementation, so the tests for unary_max will fail as well.

In light of these challenges, I think it may be more appropriate to just move these decomps back to decompositions.py, since we are unable to test the decomposition as a reference implementation. What do you think?

SS-JIA commented:
@peterbell10 apart from the location of the decomposition, it doesn't seem there are any additional discussion points. I am hoping to land this PR by EOD today to meet an internal code freeze deadline. If there are any outstanding concerns regarding the location of these decomps, let me know and I can address it in a later PR.

SS-JIA commented:
@pytorchbot merge

SS-JIA commented:
@pytorchbot merge -i "Unrelated infra CI failure. A previous version of this PR had all green signals."

SS-JIA commented:
@pytorchbot merge -i

SS-JIA commented:
Unfortunately, this diff breaks some internal CI. WIll have to revert.
